code/game/supplyshuttle.dm

```diff
@@ -688,6 +874,272 @@ var/datum/controller/supply/supply_controller = new()
 		</p></div>"
 
 	name = "[name] - [ordername]"
+
+/datum/supply_ui
+	var/atom/source_object
+	var/tgui_name = "Cargo"
+	///Id of the shuttle controlled
+	var/shuttle_id = ""
+	///Reference to the supply shuttle
+	var/obj/docking_port/mobile/supply/supply_shuttle
+	///Faction of the supply console linked
+	var/faction = FACTION_MARINE
+	///Id of the home port
+	var/home_id = ""
+
+/datum/supply_ui/New(atom/source_object)
+	. = ..()
+	src.source_object = source_object
+	RegisterSignal(source_object, COMSIG_PARENT_QDELETING, .proc/clean_ui)
+
+///Signal handler to delete the ui when the source object is deleting
+/datum/supply_ui/proc/clean_ui()
+	SIGNAL_HANDLER
+	qdel(src)
+
+/datum/supply_ui/Destroy(force, ...)
+	source_object = null
+	return ..()
+
+/datum/supply_ui/ui_host()
+	return source_object
+
+/datum/supply_ui/can_interact(mob/user)
+	. = ..()
+	if(inoperable(MAINT))
+		return UI_CLOSE
+	return TRUE
+
+/datum/supply_ui/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+
+	if(!ui)
+		if(shuttle_id)
+			supply_shuttle = SSshuttle.getShuttle(shuttle_id)
+			supply_shuttle.home_id = home_id
+			supply_shuttle.faction = faction
+		ui = new(user, src, tgui_name, source_object.name)
+		ui.open()
+
+/datum/supply_ui/ui_static_data(mob/user)
+	. = list()
+	.["categories"] = GLOB.all_supply_groups
+	.["supplypacks"] = SSpoints.supply_packs_ui
+	.["supplypackscontents"] = SSpoints.supply_packs_contents
+	.["elevator_size"] = supply_shuttle?.return_number_of_turfs()
+
+/datum/supply_ui/ui_data(mob/living/user)
```

Comment:
This function does a lot of work, would it be possible to refactor out the different parts to make it easier to read?